### PR TITLE
[PLUGIN-664] Support bulk delete in GCS Delete Action

### PR DESCRIPTION
**Pull request by JessieYee**
Jira: https://cdap.atlassian.net/browse/PLUGIN-664
PR for integration test: https://github.com/cdapio/cdap-integration-tests/pull/1180

### CODE REVIEW

File: src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java [hunk 1]
```diff
@@ -119,13 +128,77 @@ public void run(ActionContext context) throws Exception {
...
         } catch (IOException e) {
           LOG.warn(
             String.format("Failed to delete path '%s'", gcsPath)
           );
         }
       }
     }
+
+    for (GCSPath gcsPath : gcsPathsWild) {
+      String gcsPathStr = gcsPath.getUri().toString();
```
Comment (sumengwang) [inline src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java]: Can you group these stuff under the for loop into a function and write unit tests against the function. It's a little difficult to visualize the input/output.

Comment (JessieYee) [reply src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java]: Yeah, let me try to use the function

File: src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java [hunk 2]
```diff
@@ -119,13 +128,77 @@ public void run(ActionContext context) throws Exception {
       if (fs.exists(gcsPath)) {
         try {
           fs.delete(gcsPath, true);
+          deleteCount = deleteCount + 1;
```
Comment (de-lan) [inline src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java]: `deleteCount++` ?

**Review by seanzhougoogle — COMMENTED** [review]
please update the document md as well.

File: src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java (hunk 2 above)
Comment (JessieYee) [reply src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java]: done :)

File: src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java [hunk 3]
```diff
@@ -119,13 +128,80 @@ public void run(ActionContext context) throws Exception {
...
           LOG.warn(
             String.format("Failed to delete path '%s'", gcsPath)
           );
         }
       }
     }
+
+    for (GCSPath gcsPath : gcsPathsWild) {
+      String regex = getRegex(gcsPath);
+      try {
```
Comment (seanzhougoogle) [inline src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java:142]: does the logic of getting filesystem need to be in the loop ?

File: src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java [hunk 4]
```diff
@@ -119,13 +128,80 @@ public void run(ActionContext context) throws Exception {
...
+        fs = new Path(gcsPath.getUri()).getFileSystem(configuration);
+      } catch (IOException e) {
+        LOG.info("Failed deleting file " + gcsPath.getUri().getPath() + ", " + e.getMessage());
+        // no-op.
+        continue;
+      }
+      Page<Blob> blobs = storage.list(gcsPathsWild.get(0).getBucket());
```
Comment (seanzhougoogle) [inline src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java]: `gcsPathsWild.get(0)` -> gcsPath

Comment (seanzhougoogle) [reply src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java]: test when user input `gs://bucket*` what will happen here to `getBucket()`, try to handle this error case

File: src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java [hunk 5]
```diff
@@ -119,13 +128,80 @@ public void run(ActionContext context) throws Exception {
...
+      for (Blob blob : blobs.iterateAll()) {
+        Path filePath = new Path("gs://" + blob.getBucket() + "/" + blob.getName());
+        if (filePath.toString().matches(regex)) {
+          if (fs.exists(filePath)) {
+            Path rootPath = getRootMatch(filePath, regex);
```
Comment (seanzhougoogle) [inline src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java:155]: add some comments about why to get root path

File: src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java [hunk 6]
```diff
@@ -119,13 +128,80 @@ public void run(ActionContext context) throws Exception {
...
+        }
+      }
+    }
+    context.getMetrics().gauge("gc.file.delete.count", deleteCount);
+  }
+
+  public Path getRootMatch(Path filePath, String regx) {
+    Path parent = filePath;
+    String child = null;
+    while (parent.getParent() != null && parent.toString().matches(regx)) {
```
Comment (seanzhougoogle) [inline src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java]: try to simplify it by making parent as `current` and test `current.getParent().toString().matches(regx)`

File: src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java [hunk 7]
```diff
@@ -119,13 +128,80 @@ public void run(ActionContext context) throws Exception {
...
+    if (child == null) {
+      return parent;
+    } else if (parent.getParent() == null && parent.toString().matches(regx)) {
+      return parent;
+    } else {
+      return new Path(parent.toString() + "/" + child);
+    }
+  }
+
+  public String getRegex(GCSPath gcsPath) {
```
Comment (seanzhougoogle) [inline src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java]: discussed offline, looks like this method is not necessary, when matching root path we can just append a `/` to match regex

**Review by itsankit-google — COMMENTED** [review]
It would be good if we can use new `cdap-e2e-tests` framework for adding `e2e-tests` for these scenarios instead of cloud-itn since the plan is to move away from it. https://github.com/data-integrations/google-cloud/tree/develop/src/e2e-test/features/gcsdelete

cc - @rmstar, @tivv

File: src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java (hunk 4 above)
Comment (JessieYee) [reply src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java]: Might received error when use `gs://bucket*`. Try validate and received message: 

Invalid bucket name in path 'test_664_*'. Bucket name should only contain lower case alphanumeric, '-'. '_' and '.'. Please follow GCS naming convention: https://cloud.google.com/storage/docs/naming-buckets

File: src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java (hunk 3 above)
Comment (JessieYee) [reply src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java:142]: If we have delete objects from several bucket, we should get filesystem use different path. I am gonna to clarify if we should support delete from different buckets

File: src/e2e-test/resources/pluginParameters.properties [hunk 8]
```diff
@@ -108,6 +112,10 @@ pubsubDelimiter=@
 
 ## GCSDELETE-PLUGIN-PROPERTIES-START
 gcsDeleteObjectsList=testdata/GCS_RECURSIVE_TEST/recursiveFile1.csv,testdata/GCS_RECURSIVE_TEST/recursiveFile2.csv
+gcsDeleteObjectsList2=testdata/GCS_WILDCARD_TEST/wildcardFile2.csv,testdata/GCS_WILDCARD_TEST/wildcardFile3.csv
+gcsDeleteObjectsList3=testdata/GCS_WILDCARD_TEST/test
```
Comment (sumengwang) [inline src/e2e-test/resources/pluginParameters.properties:116]: `gcsWildcardPath3=testdata/GCS_WILDCARD_TEST/text*`
You have `text` above but `test` here. Is it a typo?

Comment (JessieYee) [reply src/e2e-test/resources/pluginParameters.properties:116]: oh yeah!

File: src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java [hunk 9]
```diff
@@ -119,13 +128,54 @@ public void run(ActionContext context) throws Exception {
...
             String.format("Failed to delete path '%s'", gcsPath)
           );
         }
       }
     }
+
+    for (GCSPath gcsPath : gcsPathsWild) {
+      String regex = ("\\Q" + gcsPath.getUri().toString() + "\\E").replace("*", "\\E[^/]*\\Q") + "(/.*)?";
+      try {
+        fs = new Path(gcsPathsWild.get(0).getUri()).getFileSystem(configuration);
```
Comment (seanzhougoogle) [inline src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java:143]: still using `gcsPathsWild.get(0)` ? then we can move it out of the iteration ?

File: src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java [hunk 10]
```diff
@@ -103,11 +108,15 @@ public void run(ActionContext context) throws Exception {
           String.format("Unable to access or create bucket %s. ", gcsPath.getBucket())
             + "Ensure you entered the correct bucket path and have permissions for it.", e);
       }
-      gcsPaths.add(new Path(gcsPath.getUri()));
+      if (gcsPath.getUri().toString().contains("*")) {
```
Comment (seanzhougoogle) [inline src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java:111]: I'm wondering whether we can handle it uniformly for wild and non-wild case ? non-wild case is just a special case of wild case , right ?

File: src/e2e-test/resources/testdata/GCS_WILDCARD_TEST/test/wildcardFile1.csv [hunk 11]
```diff
@@ -0,0 +1,9 @@
+id,name,yearofbirth,tablename,offset
```
Comment (seanzhougoogle) [inline src/e2e-test/resources/testdata/GCS_WILDCARD_TEST/test/wildcardFile1.csv]: do we need that much data ? for the csv file ? it's just for test deletion , right ? empty or even 1 character file will be enough , right ?

File: src/e2e-test/features/gcsdelete/GCSDelete.feature [hunk 12]
```diff
@@ -55,6 +55,63 @@ Feature: GCS Delete - Verification of GCS Delete plugin
     Then Verify the pipeline status is "Succeeded"
     Then Verify multiple objects "gcsDeleteObjectsList" deleted successfully by GCS Delete action plugin
 
+  @GCS_WILDCARD_TEST
```
Comment (seanzhougoogle) [inline src/e2e-test/features/gcsdelete/GCSDelete.feature]: nit, would it be better to call it GCS_DELETE_WILDCARD_TEST ?

File: src/e2e-test/resources/testdata/GCS_WILDCARD_TEST/test/wildcardFile1.csv (hunk 11 above)
Comment (JessieYee) [reply src/e2e-test/resources/testdata/GCS_WILDCARD_TEST/test/wildcardFile1.csv]: yeah, will use the empty file

File: src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java (hunk 10 above)
Comment (JessieYee) [reply src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java:111]: For non-wild case, we do not need to iterate Blobs (storage path list) for each gcspath. If we use the same logic for wild and non-wild case, I am worries if time for searching the filed to delete will increase?

File: src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java (hunk 9 above)
Comment (JessieYee) [reply src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSBucketDelete.java:143]: Setting another Jira to solve the delete from multiple bucket: https://cdap.atlassian.net/browse/PLUGIN-1291.

File: src/e2e-test/features/gcsdelete/GCSDelete.feature (hunk 12 above)
Comment (JessieYee) [reply src/e2e-test/features/gcsdelete/GCSDelete.feature]: ok

**Review by seanzhougoogle — APPROVED** [review]

File: src/e2e-test/java/io/cdap/plugin/gcsdelete/stepsdesign/GCSDelete.java [hunk 13]
```diff
@@ -119,6 +119,20 @@ public void verifyMultipleObjectsDeletedSuccessfullyByGCSDeleteActionPlugin(Stri
     BeforeActions.scenario.write("Multiple objects deleted successfully by GCS Delete plugin.");
   }
 
+  @Then("Verify objects {string} still exist after GCS Delete action plugin")
+  public void verifyObjectsExistByGCSDeleteActionPlugin(String commaSeparatedObjectsList)
+    throws IOException {
+    String[] objectsToDelete = PluginPropertyUtils.pluginProp(commaSeparatedObjectsList).split(",");
```
Comment (itsankit-google) [inline src/e2e-test/java/io/cdap/plugin/gcsdelete/stepsdesign/GCSDelete.java]: nit->Is it better to call them `objectsToKeep` rather than `objectsToDelete`?

Comment (JessieYee) [reply src/e2e-test/java/io/cdap/plugin/gcsdelete/stepsdesign/GCSDelete.java]: right :)

**Review by itsankit-google — APPROVED** [review]
Please squash commits before merging.
